Detectors/MUON/MCH/Tracking/src/TrackFinder.cxx

```diff
@@ -0,0 +1,1668 @@
+    nextChamber = chamber - 1;
+  } else if (chamber < lastChamber) {
+    nextChamber = chamber + 1;
+  }
+
+  // loop over all DEs of plane1
+  TrackParam paramAtCluster1{};
+  TrackParam currentParamAtCluster1{};
+  TrackParam paramAtCluster2{};
+  std::unordered_map<int, std::unordered_set<uint32_t>> newExcludedClusters{};
+  for (auto& de1 : mClusters[plane1]) {
+
+    // skip DE without cluster
+    if (de1.second == nullptr) {
+      continue;
+    }
+
+    // get the list of excluded clusters for the DE
+    auto itExcludedClusters = excludedClusters.find(de1.first);
+    bool hasExcludedClusters = (itExcludedClusters != excludedClusters.end());
+
+    // look for cluster candidate in this DE
+    for (const auto& cluster1 : *de1.second) {
+
+      // skip excluded clusters
+      if (hasExcludedClusters && itExcludedClusters->second.count(cluster1.getUniqueId()) > 0) {
+        continue;
+      }
+
+      // try to add the current cluster
+      if (!isCompatible(paramAtChamber, cluster1, paramAtCluster1)) {
+        continue;
+      }
+
+      // add it to the list of excluded clusters for this candidate
+      excludedClusters[de1.first].emplace(cluster1.getUniqueId());
+
+      // skip tracks out of limits, but after checking for overlaps
+      bool isAcceptableAtCluster1 = isAcceptable(paramAtCluster1);
+
+      // save the current parameters at cluster1, reset the propagator and add MCS effects before going to plane2
+      currentParamAtCluster1 = paramAtCluster1;
+      currentParamAtCluster1.resetPropagator();
+      TrackExtrap::addMCSEffect(&currentParamAtCluster1, SChamberThicknessInX0[chamber], -1.);
+
+      // loop over all DEs of plane2
+      bool cluster2Found(false);
+      for (auto& de2 : mClusters[plane2]) {
+
+        // skip DE without cluster
+        if (de2.second == nullptr) {
+          continue;
+        }
+
+        // skip DE that do not overlap with the DE of plane1
+        if (de2.first % 100 != (de1.first % 100 + 1) % SNDE[chamber] && de2.first % 100 != (de1.first % 100 - 1 + SNDE[chamber]) % SNDE[chamber]) {
+          continue;
+        }
+
+        // look for cluster candidate in this DE
+        for (const auto& cluster2 : *de2.second) {
+
+          // try to add the current cluster
+          if (!isCompatible(currentParamAtCluster1, cluster2, paramAtCluster2)) {
+            continue;
+          }
+
+          cluster2Found = true;
+
+          // add it to the list of excluded clusters for this candidate
+          excludedClusters[de2.first].emplace(cluster2.getUniqueId());
+
+          // skip tracks out of limits
+          if (!isAcceptableAtCluster1 || !isAcceptable(paramAtCluster2)) {
+            continue;
+          }
+
+          // continue the tracking to the next chambers and attach the 2 clusters to the new tracks if any
+          auto itNewTrack = addClustersAndFollowTrack(itTrack, paramAtCluster1, &paramAtCluster2, nextChamber, lastChamber, newExcludedClusters);
```

Comment:
why do you need to fill separately ``newExcludedClusters`` then move it to ``excludedClusters`` instead of filling it directly. This will increase the number of new/deletes in the tight loops.

Reply:
because the list of excluded clusters must be empty when starting a new branch (i.e. when adding a new cluster), in order to consider every combinations, while excludedClusters already contains the list of clusters from previously tested branches (to be excluded only when testing the case where no cluster is attached to the initial candidate on the current plane).